executor/join.go

```diff
@@ -164,6 +192,22 @@ func getJoinKey(cols []*expression.Column, row Row, vals []types.Datum, bytes []
 	return false, bytes, errors.Trace(err)
 }
 
+type joinKeyEncodeFunc func(b []byte, k types.Datum) (bool, []byte, error)
+
+func getJoinKeyFromChkRow(keyColIdx []int, keyColTypes []*types.FieldType, row chunk.Row, keyBuf []byte, encodeFunc joinKeyEncodeFunc) (bool, []byte, error) {
+	keyBuf = keyBuf[:0]
+	for i, colIdx := range keyColIdx {
+		d := row.GetDatum(colIdx, keyColTypes[i])
```

Comment:
I think we'd better not to call `row.GetDatum(colIdx, keyColTypes[i])`, it's inefficient and will be deprecated in the near future.